### ISSUE TITLE
Adding one layer with many points is surprisingly slow

### ISSUE THREAD

**Issue by andy-sweet** [napari/napari#6275]
## 🐛 Bug

Adding a points layer with many points (e.g. 10^6) is surprisingly slow on my laptop, taking ~90 seconds.

After doing some quick profiling with cProfile, `Points.symbols` is looking a little suspect. Both the napari layer setter (specifically [`coerce_symbols`](https://github.com/napari/napari/blob/62cd9ed88d254111c251faf62e19c93f03dfeae9/napari/layers/points/points.py#L840)) which takes ~25 seconds and the [vispy markers call to `set_data`](https://github.com/napari/napari/blob/62cd9ed88d254111c251faf62e19c93f03dfeae9/napari/_vispy/layers/points.py#L74) (which takes ~75s) take most of the time.

Here's a screenshot of the snakeviz viewing some [profiling_stats.txt](https://github.com/napari/napari/files/12744138/profiling_stats.txt).

![Screenshot 2023-09-27 at 4 45 35 PM](https://github.com/napari/napari/assets/2608297/29b07ca7-e5f7-44fc-a201-d76564505624)

## To Reproduce

```python
import numpy as np
import napari

rng = np.random.default_rng(seed=0)
data = rng.random(size=(1_000_000, 2)) * 1000
viewer = napari.Viewer()
viewer.add_points(data)
napari.run()
```

## Expected behavior

I'd expect 10^6 points to be a little challenging on this machine, but not especially so, and would expect to see something in less than 10 seconds at least.

## Environment

```
napari: 0.5.0a2.dev343+g787fc5f45
Platform: macOS-10.16-x86_64-i386-64bit
System: MacOS 13.5.2
Python: 3.9.6 (default, Jul 30 2021, 09:31:09)  [Clang 10.0.0 ]
Qt: 5.15.2
PyQt5: 5.15.4
NumPy: 1.25.1
SciPy: 1.10.1
Dask: 2023.6.1
VisPy: 0.13.0
magicgui: 0.7.0rc1.dev6+ga3bdc90
superqt: 0.5.3
in-n-out: 0.1.5
app-model: 0.1.4
npe2: 0.7.0

OpenGL:
  - GL version:  2.1 INTEL-20.6.4
  - MAX_TEXTURE_SIZE: 16384

Screens:
  - screen 1: resolution 1920x1080, scale 1.0

Settings path:
  - /Users/asweet/Library/Application Support/napari/napari-dev-pyqt5_cc8ffd7cdbca034eda3f67dc1b401303ece224df/settings.yaml
```

## Additional context

Coming from https://github.com/napari/napari/issues/6148 which contains a creative and interesting solution for this performance bug, but we should probably fix this is some way.

**Comment by psobolewskiPhD** [comment]
I can confirm this on my arm64 M1.
Observations: python process takes 99.9% CPU, but not more.
Memory not an issue.
And once the layer shows it's shockingly performant. Like zero issues.

**Comment by aganders3** [comment]
I think the nature of the problem in the `set_data` call may be ultimately the same. That is, we're basically checking the validity of the symbols twice.

As a quick fix we could consider a basic special case where every point has the same symbol.

**Comment by Czaki** [comment]
We even have a benchmark for this. But Benchmarks have not working for over half a year.

**Comment by psobolewskiPhD** [comment]
FYI, if you press `a` which selects points in the current layer with 1e6 points napari will hang 🤣 
Maybe it should warn or something...

**Comment by Czaki** [comment]
> FYI, if you press `a` which selects points in the current layer with 1e6 points napari will hang 🤣
> Maybe it should warn or something...

did you have sample script?

**Comment by Czaki** [comment]
Ok. 
Bug is in psygnal. EventedSet is surprising slow.

**Comment by Czaki** [comment]
Still long. Next profiling needed.

**Comment by psobolewskiPhD** [comment]
I'm going to split off the selection to a new issue, because it's not really related to the OP.

**Comment by psobolewskiPhD** [comment]
With 0.5.1 I am getting between 1.12 and 1.6 second wall time for adding the points layer with 1e6 points -- this seems ok to me?
For 1e7 points it's 9 to 13 s, so about 10x worse but not horrendous for 1e7 points?
(Just FYI with  this many,  pan and zoom are reasonable, but selection and trying to manipulate a point is pretty much impossible.)

With 0.4.19 its between 5.5 and 6.5 s well time for 1e6, I didn't bother with 1e7.
(the envs are not totally identical due to numpy2, but still)
This is on my 2020 M1, so not bleeding edge.
I'm going to mark this as probably solved?

**Comment by andy-sweet** [comment]
> I'm going to mark this as probably solved?

Yes, I think the original performance issue was fixed in #6727 and I don't like keeping issues open for general performance reasons because then they never get closed, so I'm closing this.
